packages/myst-cli/src/transforms/code.ts

```diff
@@ -83,3 +83,91 @@ export function liftCodeMetadataToBlock(session: ISession, filename: string, mda
     }
   });
 }
+
+/**
+ * Check duplicated meta tags and conflict meta tags.
+ * Separate the meta tags from tag if filter is true, otherwise just go through and process.
+ */
+export function checkMetaTags(session: ISession, tags: string[], filter: boolean): string[] {
+  const metaTagsCounter = new Map();
+  for (const action of ['hide', 'remove']) {
+    for (const target of ['cell', 'input', 'output']) {
+      metaTagsCounter.set(`${action}-${target}`, 0);
+    }
+  }
+  const check = (tag: string) => {
+    const isMetaTag = metaTagsCounter.has(tag);
+    if (isMetaTag) {
+      metaTagsCounter.set(tag, metaTagsCounter.get(tag) + 1);
+    }
+    return !isMetaTag;
+  };
+  if (filter) {
+    tags.splice(0, tags.length, ...tags.filter(check));
+  } else {
+    tags.forEach(check);
+  }
+  const validMetatags = [];
+  metaTagsCounter.forEach((value, key) => {
+    if (value >= 2) {
+      session.log.warn(`tag '${key}' is duplicated`);
+    }
+  });
+  for (const target of ['cell', 'input', 'output']) {
+    const hide = metaTagsCounter.get(`hide-${target}`) > 0;
+    const remove = metaTagsCounter.get(`remove-${target}`) > 0;
+    if (hide && remove) {
+      session.log.warn(`'hide-${target}' and 'remove-${target}' both exist`);
+      validMetatags.push(`remove-${target}`);
+    } else if (hide) {
+      validMetatags.push(`hide-${target}`);
+    } else if (remove) {
+      validMetatags.push(`remove-${target}`);
+    }
+  }
+  return validMetatags;
+}
+
+/**
+ * Traverse mdast, propagate block tags to code and output
+ */
+export function propagateBlockDataToCode(session: ISession, filename: string, mdast: Root) {
+  const blocks = selectAll('block', mdast) as GenericNode[];
+  blocks.forEach((block) => {
+    if (!block.data || !block.data.tags) return;
+    if (!Array.isArray(block.data.tags)) {
+      session.log.error(`tags in code-cell directive must be a list of strings`);
```

Comment:
I am tweaking this a bit to pass in the node and file information for better console tracking:

![image](https://github.com/executablebooks/mystjs/assets/913249/61397c12-51d2-4866-b049-9193d24398a7)